Add PaymentMethod model; Remove Card/Source polymorphism

jleclanche opened the pull request:
This is a big one boys, and it addresses #573. (Depends on #577)

**This PR is necessary work to support Stripe Sources, which Stripe JS v3, incl. Stripe Elements depend on.**

As explained in the issue, Polymorphic was a preemptive solution to a problem we didn't actually have. This PR achieves four things:
1. Create a new PaymentMethod model and turn all foreign keys to StripeSource into foreign keys to PaymentMethod. That model is an abstraction tuple: (stripe id, source type). From that PaymentMethod object you can get a Card object (and later on, a BankAccount and a Source object once we implement those properly).
2. Removes the Polymorphism on StripeSource <-> Card. Card, instead of having three layers of parenting, now has only two (Model -> StripeObject -> Card). The StripeSource model is removed, and `StripeSource` is aliased to `Card` until the next release.
3. Drops django-polymorphic dependency. (yay, honestly; it was far more trouble than worth)
4. Implements a giant 18-step handcrafted migration which handles all of this complex data and state change. 

This is by far the most painful experience I've had with dj-stripe so by god let's never do this again :P

I'm not using this in production yet (won't be for the next few days at least). But I did run it over a copy of my prod data and it works fine. Keep in mind however that it is a massive migration, which includes sql statements untested on other engines (although I believe they're ISO compatible).

It's also a big functional change. The admin kinda sucks now because there's no way to get from a source id to a target payment method (card, ...). There's some weird stuff which isnt complete yet eg. customer.sources points to a card manager, not a payment method manager; any code that touched these objects will need some kind of migration path there.

Honestly I'm positively and completely out of strength working on this. It needs a bunch of new tests, I would appreciate some help there. I'd also appreciate help ensuring that the migration works on mysql/sqlite.
I'd like to recommend an 1.1.0 release essentially immediately after this lands and we confirm it works, so that we can address #566. The amount of surface area this adds is not even funny. This also means that if there's anything backwards-incompatible you guys might want to land before 1.1.0 and/or a migration reset, now's the time. (I myself would like to land the pattern proposed in #513 before 1.1.0, so I'll try to do that).

jleclanche committed:
Move customer field from Source to Card

jleclanche committed:
Make payment method foreign keys point to an intermediary model

jleclanche committed:
Implement payment method creation and lookups

jleclanche committed:
Work around broken customer.source.deleted webhook handling

jleclanche committed:
Finish unpolymorphing the Card model

jleclanche committed:
Work around a Django bug in card migration

https://code.djangoproject.com/ticket/28573

jleclanche committed:
Make stripe_id unique only once it's backfilled

jleclanche committed:
Update tests to find PaymentMethod even when unsupported

jleclanche committed:
Remove a now-unnecessary workaround of a Django Polymorphic bug

jleclanche committed:
Drop django-polymorphic dependency

jleclanche committed:
Fix migration path and dependency

lskillen reviewed (dismissed):
:+1: This looks great to me, but I'd like to try it out against a copy of a live Production MySQL database first - It doesn't look like there is a reason for it to fail, but better safe than annoyed and angry (while manually rolling back migrations). ;-)

lskillen reviewed (dismissed):
Yep, looks good to me - Let the migration commence!

lskillen commented on djstripe/models.py:
```diff
@@ -1626,6 +1643,10 @@ class Card(StripeSource):
     )
 
+    customer = models.ForeignKey(
+        "Customer", on_delete=models.CASCADE, related_name="sources"
```
This is going to cause issues later when other payment sources also want to use the same `related_name`. This works for now, for backwards compatibility, but just wanted to note it. We may want to change how we do this sooner rather than later, then deprecate it in the next release, and then remove it (or move it to `PaymentMethod` instead).

lskillen commented on djstripe/models.py:
```diff
@@ -52,6 +51,40 @@
+    - `type` can be `card`, `bank_account` or `source`.
+    """
+    id = CharField(max_length=255, primary_key=True)
+    type = CharField(max_length=12, db_index=True)
```
There's no method here to enumerate payment sources for a customer (it has to be done via Card atm).

lskillen commented on djstripe/models.py:
```diff
@@ -52,6 +51,40 @@
+        return Card
+
+    def get_object(self):
+        return self.object_model.objects.get(stripe_id=self.id)
```
Slightly concerned about the double de-reference necessary to get the underlying, but only slightly. :-)

lskillen replied:
Might be worth adding a convenience accessor at least:

```
@cached_property
def source(self):
    return self.get_object()
```

jleclanche committed:
Make default_api_key logic available in djstripe.settings

jleclanche committed:
tests: Add event post helper for webhook tests

jleclanche committed:
tests: Send a placeholder Stripe-Signature in event creation

jleclanche committed:
Prevent DoesNotExist error when receiving multiple *.deleted events

jleclanche committed:
Explicitly call purge() on customer.deleted events

Because of the previous commit, the Customer.delete() method is no
longer called since it is not called on QuerySet.delete(). We now need
to be more explicit about calling purge().

jleclanche committed:
Fix SubscriptionRestView with Django Rest Framework 3.7.x

lskillen reviewed (approved):
LGTM - Caveats: Previous comments still stand, and should we integrate the changes @kavdev wanted to make? It could probably be done afterwards.
